File: python/thomson_multitaper/multitaper.py
"""Translation of Multitaper.m: the exact (eigenvalue-cutoff) multitaper
spectral estimator. Original Copyright 2019, Santhosh Karnik.
"""
from __future__ import annotations

import numpy as np

from ._utils import datawrap, dpss


class Multitaper:
    """Exact multitaper spectral density estimator.

    Parameters
    ----------
    n : signal length the estimator is built for
    w : half-bandwidth (as a fraction of the sampling rate)
    cutoff : if >= 1, use the first `floor(cutoff)` DPSS tapers; if in
        (eps, 1), use all tapers whose concentration eigenvalue exceeds
        `cutoff`.
    """

    def __init__(self, n: int, w: float, cutoff: float):
        self.n = n
        self.w = w
        self.cutoff = cutoff
        eps = np.finfo(float).eps

        if cutoff >= 1:
            k = int(np.floor(cutoff + eps))
            s, lam = dpss(n, n * w, k)
        elif cutoff > eps:
            if cutoff > 0.5:
                k_est = int(np.ceil(2 * n * w))
            else:
                k_est = int(np.median([1, np.ceil(2 * n * w + (2 / np.pi**2) * np.log(n) * np.log(1.0 / cutoff)) + 10, n]))
            s_full, lam_full = dpss(n, n * w, k_est)
            k = max(int(np.count_nonzero(lam_full > cutoff)), 1)
            s, lam = s_full[:, :k], lam_full[:k]
        else:
            raise ValueError("cutoff must be >= 1 or > eps")

        self.S = s
        self.lambda_ = lam
        self.K = k

    def spectral_estimate(self, x: np.ndarray, m: int | None = None) -> np.ndarray:
        """y[m] = (1/K) sum_k |FFT_M(taper_k * x)[m]|^2, the multitaper PSD estimate."""
        if m is None:
            m = self.n
        x = np.asarray(x)

        if self.n > m:
            sx = self.S * x[:, None]
            a = np.stack([datawrap(sx[:, k], m) for k in range(self.K)], axis=1)
        else:
            a = self.S * x[:, None]

        a = np.fft.fft(a, n=m, axis=0)
        a = a.real**2 + a.imag**2
        return a.sum(axis=1) / self.K

File: verification/octave_verify_multitaper/Multitaper.m
function MTSE = Multitaper(N,W,cutoff)
%% Multitaper.m 
% Function implementing the exact multitaper spectral estimator
% 
% Inputs:
% -N: Length of DPSSs
% -W: Half Bandwidth for DPSSs
% -cutoff: Projection onto the first K = cutoff DPSSs 
%          OR Projection onto the first K = #{ell : lambda_{N,W}^{(ell)} >= cutoff} DPSSs.
% Outputs:
% -MTSE an object containing an exact Multitaper Spectral Estimate function
%
% Given an Nx1 vector x, performing
% y = MTSE.SpectralEstimate(x,M);
% results in y[m] = (1/K)x^*E_(m/M)S_KS_K^*E_(m/M)^*x
% i.e. y[m] is the multitaper spectral estimate at f = m/M
%
% Notes: 
%
% Most recent change - 01/30/2019
%
% Copyright 2019, Santhosh Karnik

%% Exact Multitaper Function
MTSE.SpectralEstimate = @SpectralEstimate;

%% Store signal length, bandwidth, and number of DPSSs
MTSE.N = N;
MTSE.W = W;
MTSE.cutoff = cutoff;

%% Compute and store DPSSs
if(cutoff >= 1)
    K = floor(cutoff+eps);
    [MTSE.S,MTSE.lambda] = dpss(N,N*W,K);
elseif(cutoff > eps)
    if(cutoff > 0.5)
        Kest = ceil(2*N*W);
    else
        Kest = median([1,ceil(2*N*W+(2/pi^2)*log(N)*log(1./cutoff))+10,N]);
    end
    [S,lambda] = dpss(N,N*W,Kest);
    K = nnz(lambda > cutoff);
    K = max(K,1);
    MTSE.S = S(:,1:K);
    MTSE.lambda = lambda(1:K);
end
MTSE.K = K;

%% Exact Multitaper Spectral Estimation
function y = SpectralEstimate(x,M)
    if(nargin < 2)
        M = MTSE.N;
    end
    
    % Compute tapered signals
    if(MTSE.N > M)
       Sx = bsxfun(@times,MTSE.S,x);
       % Datawrap Sx
       A = zeros(M,MTSE.K);
       for k = 1:MTSE.K,
           A(:,k) = datawrap(Sx(:,k),M);
       end
    else
        A = bsxfun(@times,MTSE.S,x);
    end
    
    % Compute FFTs of tapered signals and sum squared magnitudes
    A = fft(A,M,1);
    A = real(A).^2+imag(A).^2;
    y = sum(A,2)/MTSE.K;
end

end